File: app/src/main/java/com/lincoln4791/dailyexpensemanager/view/MonthlyCategoryWiseDetails.java
package com.lincoln4791.dailyexpensemanager.view;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.database.Cursor;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageView;
import android.widget.TextView;

import com.lincoln4791.dailyexpensemanager.Adapters.Adapter_MonthlyCategoryWiseReport;
import com.lincoln4791.dailyexpensemanager.R;
import com.lincoln4791.dailyexpensemanager.common.Constants;
import com.lincoln4791.dailyexpensemanager.common.NodeName;
import com.lincoln4791.dailyexpensemanager.common.SQLiteHelper;
import com.lincoln4791.dailyexpensemanager.common.UtilDB;
import com.lincoln4791.dailyexpensemanager.model.MC_Posts;

import java.util.ArrayList;
import java.util.List;

public class MonthlyCategoryWiseDetails extends AppCompatActivity {

    private TextView tv_type,tv_category,tv_currentBalance_toolbar;
    private ImageView iv_home;
    private RecyclerView recyclerView;
    private List<MC_Posts> postList;
    private Adapter_MonthlyCategoryWiseReport adapter_monthlyCategoryWiseReport;
    private LinearLayoutManager linearLayoutManager;


    private String year,month,type,category;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_monthly_category_wise_details);




        //**************************************************** View Bindings ***********************************
        tv_type = findViewById(R.id.tv_type_MonthlyCategoryWise);
        tv_category = findViewById(R.id.tv_category_MonthlyCategoryWise);
        recyclerView = findViewById(R.id.rv_categoryWiseReport);

        tv_currentBalance_toolbar = findViewById(R.id.tv_currentBalanceValue_toolBar_MonthlyCategoryWiseReport);
        iv_home = findViewById(R.id.iv_home_toolbar_MonthlyCategoryWiseReport);




        //**************************************************** Initializations *********************************
        getSupportActionBar().hide();
        linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setReverseLayout(true);
        linearLayoutManager.setStackFromEnd(true);
        postList = new ArrayList<>();
        adapter_monthlyCategoryWiseReport = new Adapter_MonthlyCategoryWiseReport(postList,this);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setAdapter(adapter_monthlyCategoryWiseReport);







        //*****************************************Click Listener********************************************
        iv_home.setOnClickListener(v -> {
            startActivity(new Intent(MonthlyCategoryWiseDetails.this,MainActivity.class));
        });






        //******************************************** Starting Methods **********************************
        tv_currentBalance_toolbar.setText(String.valueOf(UtilDB.currentBalance));
        getIntentData();
        setCategoryAndType();
       new YearMonthTypeCategoryTask().execute();

    }




    private void setCategoryAndType() {
            tv_category.setText(category);
            if(type.equals(Constants.TYPE_INCOME)){
                tv_type.setText(getString(R.string.Incomes));
            }
            else if(type.equals(Constants.TYPE_EXPENSE)){
                tv_type.setText(getString(R.string.Expenses));
            }
    }





    private void getIntentData() {
       year = getIntent().getStringExtra(NodeName.POST_YEAR);
       month = getIntent().getStringExtra(NodeName.POST_MONTH);
       type = getIntent().getStringExtra(NodeName.POST_TYPE);
       category = getIntent().getStringExtra(NodeName.POST_CATEGORY);

    }







    class YearMonthTypeCategoryTask extends AsyncTask {

        @Override
        protected Object doInBackground(Object[] objects) {

            fetchYearMonthTypeCategoryWise();
            return null;
        }


        @Override
        protected void onPostExecute(Object o) {
            super.onPostExecute(o);
            Log.d("tag","listSIze "+postList.size());
            //Collections.reverse(vm_fullReport.postsList);
            recyclerView.setAdapter(adapter_monthlyCategoryWiseReport);
            adapter_monthlyCategoryWiseReport.notifyDataSetChanged();
        }
    }







    private void fetchYearMonthTypeCategoryWise() {
        postList.clear();
        //tv_typeTitle.setText(getString(R.string.Category));
        SQLiteHelper sqLiteHelper = new SQLiteHelper(this);
        Cursor cursor = sqLiteHelper.loadYearMonthTypeCategoryWise(year, month, type, category);

        while (cursor.moveToNext()) {
            String postDescription = cursor.getString(1);
            String postCategory = cursor.getString(2);
            String postType = cursor.getString(3);
            String postAmount = cursor.getString(4);
            String postTime = cursor.getString(5);
            String postDay = cursor.getString(6);
            String postMonth = cursor.getString(7);
            String postYear = cursor.getString(8);
            String postDateTime = cursor.getString(9);
            String timeStamp = cursor.getString(10);

            MC_Posts post = new MC_Posts(postDescription, postCategory, postType, postAmount, postYear, postMonth, postDay, postTime, timeStamp, postDateTime);
            postList.add(post);
        }

        cursor.close();
    }




}

File: app/src/main/java/com/lincoln4791/dailyexpensemanager/view/MonthlyCategoryWiseDetails.kt
package com.lincoln4791.dailyexpensemanager.view

import androidx.appcompat.app.AppCompatActivity
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.lincoln4791.dailyexpensemanager.model.MC_Posts
import com.lincoln4791.dailyexpensemanager.Adapters.Adapter_MonthlyCategoryWiseReport
import androidx.recyclerview.widget.LinearLayoutManager
import android.os.Bundle
import com.lincoln4791.dailyexpensemanager.R
import android.content.Intent
import com.lincoln4791.dailyexpensemanager.view.MainActivity
import com.lincoln4791.dailyexpensemanager.common.UtilDB
import com.lincoln4791.dailyexpensemanager.common.NodeName
import android.os.AsyncTask
import android.util.Log
import android.view.View
import android.widget.ImageView
import com.lincoln4791.dailyexpensemanager.common.Constants
import com.lincoln4791.dailyexpensemanager.common.SQLiteHelper
import com.lincoln4791.dailyexpensemanager.databinding.ActivityMonthlyCategoryWiseDetailsBinding
import java.util.ArrayList

class MonthlyCategoryWiseDetails : AppCompatActivity() {
    private lateinit var postList: MutableList<MC_Posts>
    private lateinit var adapter_monthlyCategoryWiseReport: Adapter_MonthlyCategoryWiseReport
    private lateinit var linearLayoutManager: LinearLayoutManager
    private lateinit var year: String
    private lateinit var month: String
    private lateinit var type: String
    private lateinit var category: String

    private lateinit var binding : ActivityMonthlyCategoryWiseDetailsBinding
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMonthlyCategoryWiseDetailsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        //**************************************************** Initializations *********************************
        supportActionBar!!.hide()
        linearLayoutManager = LinearLayoutManager(this)
        linearLayoutManager!!.reverseLayout = true
        linearLayoutManager!!.stackFromEnd = true
        postList = ArrayList()
        adapter_monthlyCategoryWiseReport = Adapter_MonthlyCategoryWiseReport(postList as ArrayList<MC_Posts>, this)
        binding.rvCategoryWiseReport.setLayoutManager(linearLayoutManager)
        binding.rvCategoryWiseReport.setAdapter(adapter_monthlyCategoryWiseReport)


        //*****************************************Click Listener********************************************
        binding.ivHomeToolbarMonthlyCategoryWiseReport.setOnClickListener(View.OnClickListener { v: View? ->
            startActivity(Intent(this@MonthlyCategoryWiseDetails,
                MainActivity::class.java))
        })


        //******************************************** Starting Methods **********************************
        binding.tvCurrentBalanceValueToolBarMonthlyCategoryWiseReport.text = UtilDB.currentBalance.toString()
        intentData
        setCategoryAndType()
        YearMonthTypeCategoryTask().execute()
    }

    private fun setCategoryAndType() {
        binding.tvCategoryMonthlyCategoryWise.text = category
        if (type == Constants.TYPE_INCOME) {
            binding.tvTypeMonthlyCategoryWise.text = getString(R.string.Incomes)
        } else if (type == Constants.TYPE_EXPENSE) {
            binding.tvTypeMonthlyCategoryWise.text = getString(R.string.Expenses)
        }
    }

    private val intentData: Unit
        get() {
            year = intent.getStringExtra(NodeName.POST_YEAR)!!
            month = intent.getStringExtra(NodeName.POST_MONTH)!!
            type = intent.getStringExtra(NodeName.POST_TYPE)!!
            category = intent.getStringExtra(NodeName.POST_CATEGORY)!!
        }

    internal inner class YearMonthTypeCategoryTask : AsyncTask<Any?, Any?, Any?>() {
        override fun doInBackground(objects: Array<Any?>): Any? {
            fetchYearMonthTypeCategoryWise()
            return null
        }

        override fun onPostExecute(o: Any?) {
            super.onPostExecute(o)
            Log.d("tag", "listSIze " + postList!!.size)
            //Collections.reverse(vm_fullReport.postsList);
            binding.rvCategoryWiseReport.adapter = adapter_monthlyCategoryWiseReport
            adapter_monthlyCategoryWiseReport!!.notifyDataSetChanged()
        }
    }

    private fun fetchYearMonthTypeCategoryWise() {
        postList!!.clear()
        //tv_typeTitle.setText(getString(R.string.Category));
        val sqLiteHelper = SQLiteHelper(this)
        val cursor = sqLiteHelper.loadYearMonthTypeCategoryWise(year, month, type, category)
        while (cursor.moveToNext()) {
            val postDescription = cursor.getString(1)
            val postCategory = cursor.getString(2)
            val postType = cursor.getString(3)
            val postAmount = cursor.getString(4)
            val postTime = cursor.getString(5)
            val postDay = cursor.getString(6)
            val postMonth = cursor.getString(7)
            val postYear = cursor.getString(8)
            val postDateTime = cursor.getString(9)
            val timeStamp = cursor.getString(10)
            val post = MC_Posts(postDescription,
                postCategory,
                postType,
                postAmount,
                postYear,
                postMonth,
                postDay,
                postTime,
                timeStamp,
                postDateTime)
            postList!!.add(post)
        }
        cursor.close()
    }
}